cupy/_core/fusion.pyx

```diff
@@ -611,17 +611,11 @@ class _FusionHistory(object):
                         out_var = make_fusion_var(out_var, ndim)
                         out_vars.append(out_var)
                         ret.append(out_var)
-                    elif numpy.can_cast(out_dtypes[i], out_vars[i].dtype,
-                                        'same_kind'):
-                        out_var = out_vars[i]
-                        ret.append(out_var)
-                    else:
-                        raise TypeError(
-                            'output (typecode \'{}\') could not be coerced '
-                            'to provided output parameter (typecode \'{}\') '
-                            'according to the casting rule '
-                            '"same_kind"'.format(
-                                out_dtypes[i].char, out_vars[i].dtype.char))
+
+                    _raise_if_invalid_cast(
+                        out_dtypes[i], out_vars[i], 'same_kind',
```

Comment:
`out_vars[i]` is not an actual `cupy.ndarray` object, but another array object with dummy values for tracing function in `cupy.fuse`.
```suggestion
                        out_dtypes[i], out_vars[i].dtype, 'same_kind',
```